src/transformers/image_utils.py

```diff
@@ -62,6 +65,190 @@
+    if callable(output) and not isinstance(output, type):
+        return _method_decorator
+
+    # Recusively wrap the output if it is a list or tuple
+    if isinstance(output, (list, tuple)):
+        return type(output)(_output_wrapper(item) for item in output)
+
+    return _cast_to_image_object(output)
+
+
+class ImageArray(np.lib.mixins.NDArrayOperatorsMixin):
+    """
+    Array container for images. This class is a wrapper around `np.ndarray` that includes image specific information
+    such as the channel dimension format and the number of channels.
+
+    See: https://numpy.org/doc/stable/user/basics.dispatch.html
+
+    Args:
+        data (`Union["PIL.Image.Image", np.ndarray, "torch.Tensor", "tf.Tensor"]`):
+            The image data.
+        data_format (`ChannelDimension`, *optional*):
+            The channel dimension format of the image. If unset, will be inferred from the image data.
+        num_channels (`int`, *optional*):
+            The number of channels in the image. If unset, will be inferred from the image data.
+    """
+
+    def __init__(
+        self,
+        data: Union["PIL.Image.Image", np.ndarray, "torch.Tensor", "tf.Tensor"],
+        data_format: Optional[ChannelDimension] = None,
+        num_channels: int = None,
+    ) -> None:
+        if isinstance(data, ImageArray):
+            self._data = data._data
+            self._data_format = data._data_format
+            self._num_channels = data._num_channels
+            self._height = data._height
+            self._width = data._width
+            return
+
+        if isinstance(data, PIL.Image.Image):
+            if data.mode in ("1", "L", "P", "I;16", "I", "F"):
+                data_format = ChannelDimension.NONE
+            else:
+                # PIL images are always channels last when converted to numpy arrays
+                data_format = ChannelDimension.LAST
+
+        data = np.array(data)
+
+        if not _can_be_image_array(data):
+            raise ValueError("Invalid image data")
+
+        if data_format is None:
+            data_format = infer_channel_dimension_format(data)
+
+        if data_format == ChannelDimension.NONE:
+            num_channels = 0
+            height = data.shape[-2]
+            width = data.shape[-1]
+        elif data_format == ChannelDimension.FIRST:
+            num_channels = data.shape[-3]
+            height = data.shape[-2]
+            width = data.shape[-1]
+        elif data_format == ChannelDimension.LAST:
+            num_channels = data.shape[-1]
+            height = data.shape[-3]
+            width = data.shape[-2]
+        else:
+            raise ValueError("Invalid channel dimension format")
+
+        self._data = data
+        self._data_format = data_format
+        self._num_channels = num_channels
+        self._height = height
+        self._width = width
+
+    def __getattribute__(self, __name: str) -> Any:
```

Comment:
I'm not very familiarized with `__getattribute__`. Sorry if this is a dummy question: Why using `__getattribute__` to expose all variables outside the class, instead of exposing them individually with `@property` ?

By removing  `__get_attribute__`, it is possible to make `PIL.Image.fromarray(img_array)` work using `__array_interface__`, with the following steps:

**Step 1:** Comment out `__get_attribute__`. 

**Step 2:** include these methods in your `ImageArray`:

```python
    @property        
    def __array_interface__(self):
        return {
                "version": 3,
                "shape": self._data.shape, 
                "typestr": self._data.dtype.str,
                "strides": self._data.strides, # Pillow takes anything, but not None
                "data": self._data.data,
                }
    
    def tobytes(self):
        return self._data.tobytes()
```
Explanation: Pillow's `from_array(obj)` method ([here](https://github.com/python-pillow/Pillow/blob/main/src/PIL/Image.py#L3012)) makes usage of `__array_interface__` (a dictionary) to convert `obj` to array without need to go down level (e.g. using C code or marshalling). So, creating the `__array_interface__` as documented [here](https://numpy.org/doc/stable/reference/arrays.interface.html) we can make our object behave like a numpy array. :nerd_face: 

I'm not sure, but I believe the problem with `__get_attribute__` is that  when it calls `_output_wrapper` with the `__array_interface__` dictionary, somehow it uses a different dictionary. 
 
**Step 3:** Now you can pass an `ImageArray` object to `PIL.Image.fromarray`, like in the example below:

```python 

import requests
import PIL

# Mocking a pillow image
url = "https://images.pexels.com/photos/2869354/pexels-photo-2869354.jpeg"
im = PIL.Image.open(requests.get(url, stream=True).raw).convert('RGB')

# Transform pillow im to ImageArray
img_array = ImageArray(im)

# Transform ImageArray object back to pillow. The `img_array` object will be seen like a numpy for PIL.Image.
pillow = PIL.Image.fromarray(img_array)
pillow.save("test.jpg")  # ---> Just to visualize that the image is "reconstructed" back to PIL.Image object.
```

`torch.from_numpy` behaves differently. I need to investigate it further.